### PR TITLE
feat(animation): apply data-driven enter animation for marks

### PR DESCRIPTION
**Pull request by pearmini**
Apply data-driven enter animation for marks.(**Test coverage: 95%**)

- **Runtime**: Apply animation for enter marks.
- **Animation**: 
  - FadeIn, ScaleInY
  - Animation is bind to shape rather than mark. (Enable to split animation and shape for Liquid.)
- **Statistic:** StackEnter
- **Shape**: Set ScaleInY as default animation for rect and hollow rect.

Basic example.

basic

```js
const options = {
  type: 'interval',
  data: [
    { genre: 'Sports', sold: 275 },
    { genre: 'Strategy', sold: 115 },
    { genre: 'Action', sold: 120 },
    { genre: 'Shooter', sold: 350 },
    { genre: 'Other', sold: 150 },
  ],
  encode: {
    x: 'genre',
    y: 'sold',
    color: 'genre',
  },
  animate: {
    enter: {
      duration: 1000, //  Specify effect time by animate options.
    },
  },
};
```

Treat enterType as channel.

enterType

```js
const options = {
  type: 'interval',
  data: [
    { genre: 'Sports', sold: 275 },
    { genre: 'Strategy', sold: 115 },
    { genre: 'Action', sold: 120 },
    { genre: 'Shooter', sold: 350 },
    { genre: 'Other', sold: 150 },
  ],
  scale: {
    enterType: { range: ['scaleInY', 'fadeIn'] }, // Specify animation types.
  },
  encode: {
    x: 'genre',
    y: 'sold',
    color: 'genre',
    // Intervals with high value will apply scaleInY animation,
    // intervals with low value will apply fadeIn animation.
    enterType: (d) => (d.sold > 200 ? 'high' : 'low'), 
  },
  animate: {
    enter: {
      duration: 3000,
    },
  },
};
```

Treat enterDuration and enterDelay as channel. ([Demo in Canis](https://canisjs.github.io/canis-editor/index.html?exmp=gantt_1))

enterDelay enterDuration

```js
const options = {
  type: 'interval',
  data: [
    { name: 'event planning', startTime: 1, endTime: 4 },
    { name: 'layout logistics', startTime: 3, endTime: 13 },
    { name: 'select vendors', startTime: 5, endTime: 8 },
    { name: 'hire venue', startTime: 9, endTime: 13 },
    { name: 'hire caterer', startTime: 10, endTime: 14 },
    { name: 'hire event decorators', startTime: 12, endTime: 17 },
    { name: 'rehearsal', startTime: 14, endTime: 16 },
    { name: 'event celebration', startTime: 17, endTime: 18 },
  ],
  coordinate: [{ type: 'transpose' }],
  scale: {
    // All the intervals will show up in 10s.
    // But the animation will take more than 10s to finish.
    enterDelay: { range: [0, 10000] }, 
  },
  encode: {
    x: 'name',
    y: ['endTime', 'startTime'],
    color: 'name',
    // The appear time of interval is linearly related to startTime.
    enterDelay: 'startTime',
    // The duration of interval animation is linearly related to duration time.
    enterDuration: (d) => d.endTime - d.startTime,
  },
};
```
Intervals show up one by one by **StackEnter**: Group marks into groups and stack enterDelay among groups. **It will work only with enterDuration channel specified.**

basicStackEnter

```js
const options = {
  type: 'interval',
  data: [
    { genre: 'Sports', sold: 275 },
    { genre: 'Strategy', sold: 115 },
    { genre: 'Action', sold: 120 },
    { genre: 'Shooter', sold: 350 },
    { genre: 'Other', sold: 150 },
  ],
  statistic: [{ type: 'stackEnter' }],
  encode: {
    x: 'genre',
    y: 'sold',
    color: 'genre',
    enterDuration: 1000,
  },
};
```
Intervals show up group by group by **StackEnter**.
groupByGroup

```js
const options = {
  type: 'interval',
  data: [
    { type: 'Apple', year: '2001', value: 260 },
    { type: 'Orange', year: '2001', value: 100 },
    { type: 'Banana', year: '2001', value: 90 },
    { type: 'Apple', year: '2002', value: 210 },
    { type: 'Orange', year: '2002', value: 150 },
    { type: 'Banana', year: '2002', value: 30 },
  ],
  // Group intervals by x channel which means grouping by year field.
  statistic: [{ type: 'stackEnter', by: ['x'] }],
  encode: {
    x: 'year',
    y: 'value',
    color: 'type',
    series: 'type',
    enterDuration: 1000,
  },
};
```
Intervals show up series by series by **StackEnter**.

seriesBySeries

```js
const options = {
  type: 'interval',
  data: [
    { type: 'Apple', year: '2001', value: 260 },
    { type: 'Orange', year: '2001', value: 100 },
    { type: 'Banana', year: '2001', value: 90 },
    { type: 'Apple', year: '2002', value: 210 },
    { type: 'Orange', year: '2002', value: 150 },
    { type: 'Banana', year: '2002', value: 30 },
  ],
  statistic: [{ type: 'stackEnter', by: ['color'] }],
  encode: {
    x: 'year',
    y: 'value',
    color: 'type',
    series: 'type',
    enterDuration: 1000,
  },
};
```

Intervals show up group by group then series by series by **StackEnter**. ([Demo in Canis](https://canisjs.github.io/canis-editor/index.html?exmp=groupedBar_1))

groupByGroupSeriesBySeries

```js
const options = {
  type: 'interval',
  data: [
    { type: 'Apple', year: '2001', value: 260 },
    { type: 'Orange', year: '2001', value: 100 },
    { type: 'Banana', year: '2001', value: 90 },
    { type: 'Apple', year: '2002', value: 210 },
    { type: 'Orange', year: '2002', value: 150 },
    { type: 'Banana', year: '2002', value: 30 },
  ],
  statistic: [{ type: 'stackEnter', by: ['color', 'x'] }],
  encode: {
    x: 'year',
    y: 'value',
    color: 'type',
    series: 'type',
    enterDuration: 1000,
  },
};
```
Intervals show up series by series then group by group by **StackEnter**.
seriesBySeriesGroupByGroup

```js
const options = {
  type: 'interval',
  data: [
    { type: 'Apple', year: '2001', value: 260 },
    { type: 'Orange', year: '2001', value: 100 },
    { type: 'Banana', year: '2001', value: 90 },
    { type: 'Apple', year: '2002', value: 210 },
    { type: 'Orange', year: '2002', value: 150 },
    { type: 'Banana', year: '2002', value: 30 },
  ],
  statistic: [{ type: 'stackEnter', by: ['x', 'color'] }],
  encode: {
    x: 'year',
    y: 'value',
    color: 'type',
    series: 'type',
    enterDuration: 1000,
  },
};
```

StackEnter is useful for stacked intervals. StackEnter maybe inferred for stack intervals with scaleInY animation in the future work.

| Without StackEnter      |With StackEnter|
| ----------- | ----------- |
|    stackBar  |    stackBarWithStackEnter |

```js
const options = {
  type: 'interval',
  data: [
    { type: 'Apple', year: '2001', value: 260 },
    { type: 'Orange', year: '2001', value: 100 },
    { type: 'Banana', year: '2001', value: 90 },
    { type: 'Apple', year: '2002', value: 210 },
    { type: 'Orange', year: '2002', value: 150 },
    { type: 'Banana', year: '2002', value: 30 },
  ],
  statistic: [{ type: 'stackEnter', by: ['color'] }],
  encode: {
    x: 'year',
    y: 'value',
    color: 'type',
    enterDuration: 1000,
  },
}
```

### CODE REVIEW

**Comment by visiky** [comment]
image

一定要用 `high` `low` 来代表吗

**Comment by pearmini** [comment]
> image
> 
> 一定要用 `high` `low` 来代表吗

不一定要用 high、low，只要是两个不同的值就可以，或者直接绑定一个字段也是可以的。如果 range 少于两个，比如只有一个，那么就都用一种动画。如果 range 大于两个，就用前两个动画。

File: src/runtime/types/component.ts [hunk 1]
```diff
@@ -172,3 +179,14 @@ export type GuideComponentProps = {
 };
 export type GuideComponentComponent<O = Record<string, unknown>> =
   G2BaseComponent<GuideComponent, O, GuideComponentProps>;
+
+export type Animation = (
+  shape: DisplayObject,
+  style: Record<string, any>,
+  coordinate: Coordinate,
+  theme: G2Theme,
+) => GAnimation;
```
Comment (visiky) [inline src/runtime/types/component.ts:188]: 我会倾向于和之前的保持差不多一致?
```ts
// before
type A = (shape: IShape | IGroup, animateCfg: GAnimateCfg, cfg: AnimateExtraCfg)  => any;

// now
type Animation = (
  shape: DisplayObject,
  // 估计也都是 G 支持配置的动画属性吧?
  animateCfg: Record<string, any>,
 // coordinate 必选，可以单独出来第三个参数（但后续不知道会不会有其他的，似乎考虑第四个参数是一个 options?  目前有 theme，后续可以扩充其他的？否则参数继续在后面加的话，就会超过 4 个了
  coordinate: Coordinate,
  theme: G2Theme,
) => GAnimation;


```

File: src/runtime/types/component.ts [hunk 2]
```diff
@@ -147,9 +150,13 @@ export type Shape = (
   },
   coordinate: Coordinate,
 ) => DisplayObject;
+export type ShapeProps = {
```
[the remark this answers is not in the document]
Comment (pearmini) [reply src/runtime/types/component.ts:153]: 嗯嗯，可以，这里不空只是因为它们是相关的，我想来点“段落感”😝

File: __tests__/unit/theme/light.spec.ts [hunk 3]
```diff
@@ -6,14 +6,15 @@ describe('light', () => {
...
     });
   });
 
   it('Light({...}) overrides expected defaults', () => {
-    expect(Light({ defaultColor: 'red' })).toEqual({
-      defaultColor: 'red',
-      defaultCategory10: 'category10',
-      defaultCategory20: 'category20',
-    });
+    expect(Light({ defaultColor: 'red' }).defaultColor).toBe('red');
```
[the remark this answers is not in the document]
Comment (pearmini) [reply __tests__/unit/theme/light.spec.ts:18]: 学到了，不过现在这种写法以后也不需要每次改了，因为只是覆盖的 defaultColor，只对这个属性进行断言就好了

File: src/statistic/stackEnter.ts [hunk 4]
```diff
@@ -0,0 +1,69 @@
+import { group, max } from 'd3-array';
```
[the remark this answers is not in the document]
Comment (pearmini) [reply src/statistic/stackEnter.ts:1]: API 设计上还是有点区别，我感觉 d3 的更好用🤣

File: src/runtime/types/component.ts (hunk 1 above)
Comment (pearmini) [reply src/runtime/types/component.ts:188]: options 是动画组件的对应函数的参数，可以直接获得：

```js
function Animate(options) {
  return (shape, style, coordinate, theme) => {/*...*/ }
}
```

我感觉没有必要，其实动画和 mark 一样，都应该考虑 coordinate，所以 API 肯定会变。现在这样设计也是为了和当前 mark 等 API 保持一致性。

File: __tests__/unit/statistic/stackEnter.spec.ts [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    expect(s2(v)).toEqual({
+      index: [0, 1, 2, 3, 4, 5],
+      value: {
+        x: [[0], [0], [0], [1], [1], [1]],
+        color: ['red', 'yellow', 'blue', 'red', 'yellow', 'blue'],
+        enterDuration: [100, 100, 100, 100, 100, 100],
+        enterDelay: [0, 200, 400, 100, 300, 500],
+      },
+    });
+  });
```
[the remark this answers is not in the document]
Comment (pearmini) [reply __tests__/unit/statistic/stackEnter.spec.ts:66]: ok

**Review by visiky — APPROVED** [review]
LG;

**Pull request merged**